gtsfm/bundle/bundle_adjustment_hilti.py

```diff
@@ -0,0 +1,189 @@
+"""Factor-graph based formulation of Bundle adjustment and optimization.
+
+Authors: Xiaolong Wu, John Lambert, Ayush Baid
+"""
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple
+
+import gtsam
+from gtsam import (
+    BetweenFactorPose3,
+    NonlinearFactorGraph,
+    Pose3,
+    PriorFactorPose3,
+    Values,
+    symbol_shorthand,
+)
+
+import gtsfm.common.types as gtsfm_types
+import gtsfm.utils.logger as logger_utils
+from gtsfm.bundle.bundle_adjustment import BundleAdjustmentOptimizer
+from gtsfm.common.gtsfm_data import GtsfmData
+from gtsfm.common.pose_prior import PosePrior, PosePriorType
+from gtsfm.loader.hilti_loader import HiltiLoader
+
+METRICS_GROUP = "bundle_adjustment_metrics"
+
+METRICS_PATH: Path = Path(__file__).resolve().parent.parent.parent / "result_metrics"
+
+TEST_DATA_ROOT: Path = Path(__file__).resolve().parent.parent.parent / "tests" / "data"
+HILTI_TEST_DATA_PATH: Path = TEST_DATA_ROOT / "hilti_exp4_small"
+
+"""In this file, we use the GTSAM's GeneralSFMFactor2 instead of GeneralSFMFactor because Factor2 enables decoupling
+of the camera pose and the camera intrinsics, and hence gives an option to share the intrinsics between cameras.
+"""
+
+P = symbol_shorthand.P  # 3d point
+X = symbol_shorthand.X  # camera pose
+K = symbol_shorthand.K  # calibration
+B = symbol_shorthand.B  # body frame (IMU) pose
+
+CAM_POSE3_DOF = 6  # 6 dof for pose of camera
+CAM_CAL3BUNDLER_DOF = 3  # 3 dof for f, k1, k2 for intrinsics of camera
+CAM_CAL3FISHEYE_DOF = 9
+IMG_MEASUREMENT_DIM = 2  # 2d measurements (u,v) have 2 dof
+POINT3_DOF = 3  # 3d points have 3 dof
+
+
+# noise model params
+CAM_POSE3_PRIOR_NOISE_SIGMA = 0.1
+CAM_CAL3BUNDLER_PRIOR_NOISE_SIGMA = 1e-5  # essentially fixed
+CAM_CAL3FISHEYE_PRIOR_NOISE_SIGMA = 1e-5  # essentially fixed
+MEASUREMENT_NOISE_SIGMA = 1.0  # in pixels
+CAM_IMU_POSE_PRIOR_SIGMA = 1e-3
+
+logger = logger_utils.get_logger()
+
+
+class BundleAdjustmentHiltiOptimizer(BundleAdjustmentOptimizer):
+    """Bundle adjustment using factor-graphs in GTSAM.
+
+    This class refines global pose estimates and intrinsics of cameras, and also refines 3D point cloud structure given
+    tracks from triangulation."""
+
+    # TODO(Ayush): share calibrations between same cams?
+    # TODO(Ayush): use a diagonal model for calibration prior as distortion should have much lower sigmas?
+
+    def __init__(
+        self,
+        output_reproj_error_thresh: Optional[float] = None,
+        robust_measurement_noise: bool = False,
+        shared_calib: bool = False,
+        max_iterations: Optional[int] = None,
+    ) -> None:
+        super().__init__(output_reproj_error_thresh, robust_measurement_noise, shared_calib, max_iterations)
+
+        # Temporary hack to get relative pose between cameras and IMUs
```

Comment:
Do you plan on passing these as inputs?

Reply:
I think we need a body_T_cam: List[PosePrior] ? The length of the list would provide the number of cameras in the rig. So we would not have to hard code it.

Reply:
Yeah that would be optimal, but I guess it requires a little more work.

Reply:
And some unknowns too. For example, we will have prior on relative poses of cameras (which we use in 2-view BA) and on relative poses of imu (which we use here)